fedWireMessage.go

```diff
@@ -1793,267 +1725,334 @@ func (fwm *FEDWireMessage) isFIIntermediaryFIAdviceValid() error {
+// If present, FIBeneficiaryAdvice and Beneficiary are required.
+func (fwm *FEDWireMessage) validateFIPaymentMethodToBeneficiary() error {
 	if fwm.FIPaymentMethodToBeneficiary != nil {
-		if fwm.FIBeneficiary == nil {
-			return fieldError("FIBeneficiary", ErrFieldRequired)
+		if fwm.FIBeneficiaryAdvice == nil {
+			return fieldError("FIBeneficiaryAdvice", ErrFieldRequired)
 		}
 		if fwm.Beneficiary == nil {
 			return fieldError("Beneficiary", ErrFieldRequired)
 		}
+		return nil
 	}
 	return nil
 }
 
-func (fwm *FEDWireMessage) isUnstructuredAddendaValid() error {
-	if fwm.UnstructuredAddenda != nil {
-		if fwm.LocalInstrument != nil {
-			switch fwm.LocalInstrument.LocalInstrumentCode {
-			case
-				SequenceBCoverPaymentStructured,
-				ProprietaryLocalInstrumentCode,
-				RemittanceInformationStructured,
-				RelatedRemittanceInformation:
-				if len(fwm.UnstructuredAddenda.Addenda) > 1 {
-					return NewErrInvalidPropertyForProperty("UnstructuredAddenda", fwm.UnstructuredAddenda.String(),
-						"LocalInstrumentCode", fwm.LocalInstrument.LocalInstrumentCode)
-				}
+// validateUnstructuredAddenda validates TagUnstructuredAddenda within a FEDWireMessage
+// * Must be present if BusinessFunctionCode is CustomerTransferPlus and LocalInstrument is ANSIX12format,
+//    GeneralXMLformat, ISO20022XMLformat, NarrativeText, STP820format, SWIFTfield70 or UNEDIFACTformat;
+//    otherwise not permitted.
+// * If LocalInstrument is ANSIX12format or STP820format, only the X12 Character Set* is permitted in
+//    Addenda Information element.
+// * If LocalInstrument is GeneralXMLformat, ISO20022XMLformat, NarrativeText, SWIFTfield70 or
+//    UNEDIFACTformat, only the SWIFT MX ISO 20022 Character Set* is permitted in Addenda Information
+//    element.
+func (fwm *FEDWireMessage) validateUnstructuredAddenda() error {
+	if fwm.BusinessFunctionCode.BusinessFunctionCode == CustomerTransferPlus && fwm.LocalInstrument != nil {
+		switch fwm.LocalInstrument.LocalInstrumentCode {
+		case ANSIX12format, GeneralXMLformat, ISO20022XMLformat, NarrativeText, STP820format, SWIFTfield70, UNEDIFACTformat:
+			if fwm.UnstructuredAddenda == nil {
+				return fieldError("UnstructuredAddenda", ErrFieldRequired)
+			}
+		default:
+			if fwm.UnstructuredAddenda != nil {
+				return NewErrInvalidPropertyForProperty("UnstructuredAddenda", fwm.UnstructuredAddenda.String(),
+					"LocalInstrumentCode", fwm.LocalInstrument.LocalInstrumentCode)
 			}
 		}
+	} else {
+		if fwm.UnstructuredAddenda != nil {
+			return fieldError("UnstructuredAddenda", ErrNotPermitted)
+		}
 	}
+
+	// TODO: if LocalInstrument is ANSIX12format or STP820format, make sure Addenda Information only contains charaters within the X12 character set
+	// TODO: if LocalInstrument is any of the other permitted formats, make sure Addenda Information only contains charaters within the SWIFT MX ISO 20022 character set
```

Comment:
Should we make issues for these?

Reply:
Probably so. It's similar to this issue: https://github.com/moov-io/wire/issues/12

Reply:
Actually, re-reading issue #12 I realized it's pointing to the same requirement. No need to create separate issues.